core/modules/rproc/InfileMerger.cc

```diff
@@ -102,82 +102,48 @@ const char JOB_ID_BASE_NAME[] = "jobId";
 
 namespace {
 
-class SemaMgr {
-public:
-    explicit SemaMgr(int max) : _max(max) { assert(_max> 1); }
-    // &&& delete default constructor, copy constructor, and such
-
-    int getTotalCount() { return _totalCount; }
-    int getUsedCount() { return _usedCount; }
-    friend class SemaLock;
-
-private:
-    void _take()  {
-        ++_totalCount;
-        if (_usedCount >= _max) {
-            std::unique_lock<std::mutex> uLock(_mtx);
-            _tCv.wait(uLock, [this](){
-                return (_usedCount < _max);
-            });
-        }
-
-        // Incrementing outside the mutex may result in occasionally having more than
-        // _maxTransmits happening at one time. This should be harmless.
-        ++_usedCount;
-    }
-
-    void _release() {
-        --_totalCount;
-        --_usedCount;
-        _tCv.notify_one();
-    }
-
-    std::atomic<int> _totalCount{0};
-    std::atomic<int> _usedCount{0};
-    int _max;
-    std::mutex _mtx;
-    std::condition_variable _tCv;
-};
 
 
-/// RAII class to support SemaMgr
-class SemaLock {
-public:
-    SemaLock(SemaMgr& semaMgr)
-      : _semaMgr(semaMgr) {
-        _semaMgr._take();
-    }
-    // &&& delete default constructor and such
-
-    ~SemaLock() {
-        _semaMgr._release();
-    }
-
-private:
-    SemaMgr& _semaMgr;
-};
-
 } // anon namespace
 
 
 namespace lsst {
 namespace qserv {
 namespace rproc {
 
-
 ////////////////////////////////////////////////////////////////////////
 // InfileMerger public
 ////////////////////////////////////////////////////////////////////////
 InfileMerger::InfileMerger(InfileMergerConfig const& c,
-                           std::shared_ptr<qproc::DatabaseModels> const& dm)
+                           std::shared_ptr<qproc::DatabaseModels> const& dm,
+                           util::SemaMgr::Ptr const& semaMgrConn)
     : _config(c),
       _mysqlConn(_config.mySqlConfig),
       _databaseModels(dm),
-      _jobIdColName(JOB_ID_BASE_NAME) {
-    //_mysqlConn.closeMySqlConn(); // &&&
-    _fixupTargetName();
-    _maxResultTableSizeMB = _config.mySqlConfig.maxTableSizeMB;
+      _jobIdColName(JOB_ID_BASE_NAME),
+      _semaMgrConn(semaMgrConn) {
 
+    _fixupTargetName();
+    _maxResultTableSizeMB = _config.czarConfig.getMaxTableSizeMB();
+    _setEngineFromStr(_config.czarConfig.getResultEngine());
+    if (_dbEngine == MYISAM) {
+        LOGS(_log, LOG_LVL_INFO, "Engine is MYISAM, serial");
+        if (!_setupConnectionMyIsam()) {
+            throw InfileMergerError(util::ErrorCode::MYSQLCONNECT, "InfileMerger mysql connect failure.");
+        }
+    } else {
+        if (_dbEngine == INNODB)  {
+            LOGS(_log, LOG_LVL_INFO, "Engine is INNODB, parallel, semaMgrConn=" << _semaMgrConn);
+            // TODO: Size calculation does not work correctly for INNODB
+            //       increase max allowed size to reduce false positives
+            //       until a reasonably accurate method is implemented.
```

Comment:
Yes, that would be the right ting to do - count bytes as they're arriving from workers rather than measuring the size of the result table.